lib/rtti.h

```diff
@@ -0,0 +1,181 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef LIB_RTTI_H_
+#define LIB_RTTI_H_
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <type_traits>
+
+namespace RTTI {
+
+using TypeId = std::uint64_t;
+static constexpr TypeId InvalidTypeId = UINT64_C(0);
+
+namespace detail {
+// Apparently string_view does not optimize properly in constexpr context
+// for gcc < 13
+struct TypeNameHolder {
+    template <size_t N>
+    constexpr TypeNameHolder(const char (&str)[N]) : str(&str[0]), length(N - 1) {}  // NOLINT
+
+    const char *str;
+    size_t length;
+};
+
+template <typename T>
+constexpr TypeNameHolder TypeNameProxy() {
+#ifdef __clang__
+    return __PRETTY_FUNCTION__;
+#elif defined(__GNUC__)
+    return __PRETTY_FUNCTION__;
+#else
+#error "Unsupported compiler"
+#endif
+}
+
+static constexpr uint64_t FNV1a(const char *str, size_t n,
```

Comment:
Would it be possible to just compare the strings instead of calculating a hash? The comparison happens all at compile-time anyway if I'm looking correctly. If we ever get a collision that will be a terrible mess to debug.

Reply:
The comparison happens in runtime certainly as typeid value is used to determine the dynamic type of object pointer. Let's check `node->is<T>()` implementation for example:

```c++
    template <typename T>
    [[nodiscard]] bool is() const noexcept {
        return isA(TypeInfo<T>::id());
    }
```

Here we are deriving typeid of `T` and checking if `this` is-a given a typeid. Now, `isA` implementation is as follows (note that the function is virtual as we need to know the exact type:
```c++
    [[nodiscard]] bool isA(RTTI::TypeId typeId) const noexcept override {
        return TypeInfo::isA(typeId);
    }
```

Here `TypeInfo = RTTI::TypeInfo<T, AllBasesOfT...>`, so we can check if typeid of input type `T` is compatible with typeid of `this` and all its immediate bases. Then we perform recursion up hierarchy:
```
    [[nodiscard]] static constexpr bool isA(TypeId typeId) noexcept {
        return (id() == typeId) || (... || (Parents::TypeInfo::isA(typeId)));
    }
```

So, the check is performed in runtime as we need to know the exact child via virtual dispatch.

Side note: fold in `isA` is inlined by all the compilers I checked, so it results in a series of fine-grained comparisons (that are typically optimized and combined quite nice).

Reply:
Ah, and string comparison in runtime is not quite an option here. `__platform_strcmp` on the profile in https://github.com/p4lang/p4c/pull/4377#issuecomment-1917495763 actually comes from typeid comparison

Reply:
What about using `std::type_index`? I believe that is guaranteed to be unique, no?

Reply:
> What about using std::type_index? I believe that is guaranteed to be unique, no?

Well, it depends on how you'd define "unique". `std::type_index` is just a thin wrapper over the `type_info` pointer. It has various explicit instantiations defined (e.g. `std::hash<type_index>`), so you can use it in associate containers. Everything else (e.g. `==`, etc.) is delegated to underlying type_info. Here is one implementation to be precise: https://github.com/llvm/llvm-project/blob/main/libcxx/include/typeindex#L64 So it's unique in terms of `operator==` yes, but there are interesting details here.

We are returning essentially to the same set of problems we are trying to solve:
  - We'd need to use `typeid()` that we're trying to avoid
  - type_info comparison (`type_info::operator==` and corresponding `type_index::operator==` that just delegates to it) could be slow on some platforms and in some cases (Apple ARM64 is one of the most notable example among the others), so we'd need to resort to string comparison. Addresses of `type_info`'s are not unique.

You can check the details of Itanium C++ ABI for the description and variants of `type_info` layout, but p4c case is essentially the most complex one due to multiple virtual bases and diamond-shaped inheritance.

Another IMO important point is that this implementation allows for plug-in manual typeid that could be used for various static polymorphism as shown in https://github.com/p4lang/p4c/pull/4377#discussion_r1473533876 I do not think there is sane way of collecting typeid's from the class hierarchy via ordinary C++ RTTI.

PS: I must probably say that having custom RTTI for closed or semi-open class hierarchies is quite common thing:
  - LLVM has it. And things that are using it down the stack (Swift compiler, MLIR-based things, etc.), see https://llvm.org/docs/HowToSetUpLLVMStyleRTTI.html and https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Support/Casting.h
  - Game engines have RTTI for game objects. Here is an example RTTI implementation in Lumberyard: https://github.com/aws/lumberyard/blob/master/dev/Code/Framework/AzCore/AzCore/RTTI/RTTI.h